qiskit/circuit/library/n_local/qaoa_ansatz.py

```diff
@@ -0,0 +1,235 @@
+        from qiskit.opflow import CircuitStateFn
+        from qiskit.opflow import H, CircuitOp, EvolutionFactory
+
+        # initialize circuit, possibly based on given register/initial state
+        if isinstance(self._initial_state, QuantumCircuit):
+            circuit_op = CircuitStateFn(self._initial_state)
+        else:
+            circuit_op = (H ^ self._num_qubits)
+
+        # iterate over layers
+        for idx in range(self._reps):
+            # the first [:self._p] parameters are used for the cost operator,
+            # so we apply them here
+            circuit_op = (self._cost_operator * parameters[idx]).exp_i().compose(circuit_op)
+            from qiskit.opflow import OperatorBase
+            if isinstance(self._mixer, OperatorBase):
+                mixer = cast(OperatorBase, self._mixer)
+                # we apply beta parameter in case of operator based mixer.
+                circuit_op = (mixer * parameters[idx + self._reps]).exp_i().compose(circuit_op)
+            else:
+                # mixer as a quantum circuit that can be parameterized
+                mixer = cast(QuantumCircuit, self._mixer)
+                num_params = mixer.num_parameters
+                # the remaining [self._p:] parameters are used for the mixer,
+                # there may be multiple layers, so parameters are grouped by layers.
+                param_values = parameters[self._reps + num_params * idx:
+                                          self._reps + num_params * (idx + 1)]
+                param_dict = dict(zip(mixer.parameters, param_values))
+                mixer = mixer.assign_parameters(param_dict)
+                circuit_op = CircuitOp(mixer).compose(circuit_op)
+
+        evolution = EvolutionFactory.build(self._cost_operator)
+        circuit_op = evolution.convert(circuit_op)
+        return circuit_op.to_circuit()
+
+    @property
+    def cost_operator(self) -> 'OperatorBase':
+        """Returns an operator representing the cost of the optimization problem."""
+        return self._cost_operator
+
+    @cost_operator.setter
+    def cost_operator(self, cost_operator: 'OperatorBase') -> None:
+        """Sets cost operator."""
+        self._cost_operator = cost_operator
+        self._invalidate()
+
+    @property
+    def reps(self) -> int:
+        """Returns the `p` parameter, which determines the depth of the circuit."""
+        return self._reps
+
+    @reps.setter
+    def reps(self, reps: int) -> None:
+        """Sets the `p` parameter."""
+        self._reps = reps
+        self._invalidate()
+
+    @property
+    def initial_state(self) -> Optional[QuantumCircuit]:
+        """Returns an optional initial state as a circuit"""
+        return self._initial_state
+
+    @initial_state.setter
+    def initial_state(self, initial_state: Optional[QuantumCircuit]) -> None:
+        """Sets initial state."""
+        self._initial_state = initial_state
+        self._invalidate()
```

Comment:
@Cryoris This is an example of when None is set and None is returned rather than the layer of Hadmards that will actually be used as the initial state. Making this remark in regards of what I noted to you in regards of Grover behavior (and you said AE does this too). So here should the setter detect None and set the initial state to the Hadmards so that the getter returns that back or the getter do it the test etc. Internally then the setup of initial state would not need the if test and building of Hadmards at the point it is done. Just thinking about consistency of behavior in what we expect over the setter/getter when there is some default for None that is internally built

Reply:
True, @adekusar-drl what do you think of changing the behaviour to be the same as in QAE (and soon Grover), where the default value is returned by the getter, instead of doing it upon execution? More concretely, to do
```python
@property
def initial_state(self):
    if self._initial_state is None:
        initial = QuantumCircuit(self.num_qubits)
        initial.h(range(self.num_qubits))
        return initial
    return self._initial_state
    
# in the main part of the code ...
    circuit = self.initial_state
```
instead of (how it is currently)
```python
@property
def initial_state(self):
    return self._initial_state
    
# in the main part of the code ...
    if self._initial_state is None:
        circuit = QuantumCircuit(self.num_qubits)
        circuit.h(range(self.num_qubits))
    else:
        circuit = self._initial_state
```
This shifts the responsibility of creating defaults to the getter and keeps the main part of the code a little more organized. Also users can more easily check what the default values are.

Reply:
@Cryoris I'm not in favor of placing any logic into properties, they should be as simple as possible in my opinion. If I add this default behavior to the `initial_state` getter, that means I have to remember I must access the initial value only through the property not the corresponding field, so if I get it via `_initial_state`, say by mistake, I may get `None` value. Also, I prefer not to instantiate anything until it is really required, lazy instantiation. There could be a situation when a getter is called (causing creation of a circuit) and then a corresponding setter is called overwriting the default value created earlier.

Long story short, I can this, seems this a common approach.

Reply:
> Also, I prefer not to instantiate anything until it is really required, lazy instantiation. 

Right, but how does letting the property return the default instantiate anything before it's needed? Since we only call the getter once we need the initial state this is still lazy instantiation or not?

>  If I add this default behavior to the initial_state getter, that means I have to remember I must access the initial value only through the property not the corresponding field, so if I get it via _initial_state, say by mistake, I may get None value.

That's true. But in the current code there's no way to get the default initial state w/o constructing the whole circuit. If we want to allow getting the default and the input value we'd have to add two properties -- but doing that for all input values might be a lot and not very nice therefore... alternatively you could just say `initial_state` gives the default and `_initial_state` is what the user gave us. I'd say we should simplify access to what we need more and what the user doesn't know, hence returning the default and not the input value. 🙂 

But I don't have a super strong opinion on that, if you have a strong opposition I'm also fine not returning the default.

Reply:
I am going to jump back in and say consistency of behavior is helpful to users. I started out by noting AE and Grover have this behavior now of calling the setter to set None and the getter gives back a non-None entity (the default it internally creates). In general, as we used to do it, if you passed in None you got None back - you knew it was set to whatever the internal default was but you had direct way of just getting that default. I can see both arguments but it would be nice to settle for one way, for consistency, and ensure going forwards we do this (refactoring where it does not when we can too).

Reply:
> > Also, I prefer not to instantiate anything until it is really required, lazy instantiation.
> 
> Right, but how does letting the property return the default instantiate anything before it's needed? Since we only call the getter once we need the initial state this is still lazy instantiation or not?
> 

Now you are right, but this is a public property, you never know who's going to access this property and when, also we don't know how the code will evolve.

> > If I add this default behavior to the initial_state getter, that means I have to remember I must access the initial value only through the property not the corresponding field, so if I get it via _initial_state, say by mistake, I may get None value.
> 
> That's true. But in the current code there's no way to get the default initial state w/o constructing the whole circuit. If we want to allow getting the default and the input value we'd have to add two properties -- but doing that for all input values might be a lot and not very nice therefore... alternatively you could just say `initial_state` gives the default and `_initial_state` is what the user gave us. I'd say we should simplify access to what we need more and what the user doesn't know, hence returning the default and not the input value. 🙂
> 
 Two properties, of course, don't make sense. I added a line in the docs, that if initial state is not specified, then a set of H gates is applied, thanks to Steve for spotting this. So, I don't think it is a problem. 

@Cryoris @woodsp-ibm 
Anyway, I have implemented default values in getters. I guess that's it for now?

Reply:
Great, thx